### DISCUSSION TITLE
Best way to create custom inputs in v7: expose register?

### DISCUSSION THREAD

**Discussion by Aryaman73** [react-hook-form/react-hook-form#6430] [Q&A]
We're creating custom inputs for our application, and I was wondering whether the register command should be placed at the input level, or if it should be _exposed_ (similar to [this](https://codesandbox.io/s/react-hook-form-custom-input-b5lll?file=/src/index.js:515-523), but I think this is pre-v7). I think both can work but I was wondering what is recommended by convention:

In other words, should `MyCustomTextInput` look like this:

` <MyCustomTextInput type="red" myName="text-input-1" myMaxLength={5} />
`

where the component definition would have something like `<input {...register(myName, {maxLength: {myMaxLength}})} />`

OR, something like this:

` <MyCustomTextInput type="red" {...register(myName, {maxLength: {myMaxLength}})}  />
`

(Not sure if Controller comes into play here. I know controllers are useful if you're using MaterialUI inputs that don't expose their refs, but I wanted to know if that's how we should also approach custom inputs that we're creating ourselves)

**Comment by bluebill1049** [comment]
- forward input ref
- expose the following props 
  - onChange
  - onBlur
  - name

**Reply by Aryaman73** [reply]
By "forward input ref" are you referring to the first method?

I also understand your point about exposing onChange & onBlur, but what about validation (maxLength, minLength, validate etc.)? I know we can expose all of them individually, but could we also expose the entire "`registerOptions`" object? So something like:

`<MyCustomTextInput type="red" myName="text-input-1" registerOptions={{ maxLength: 5 }} />`

where the component definition would have something like `<input {...register(myName, registerOptions} />`

**Reply by bluebill1049** [reply]
https://reactjs.org/docs/forwarding-refs.html
